### PR TITLE
Use IPropertyNameGenerator to determine property names of complex que…

### PR DESCRIPTION
**Pull request by sweetlandj**
This is a fix for issue #1591.

It extends `ParameterModelBase` by changing the return type of `PropertyNames` and `CollectionPropertyNames` from `IEnumerable<string>` to `IEnumerable<PropertyModel>`, a new class that exposes both the original `JsonProperty` key (the schema property name) and a generated property name suitable for use as a C# or TypeScript identifier.  `PropertyModel` also contains an overridden `ToString` method that will return the generated property name, which _should_ preserve backward compatibility with existing usages of `{{property}}` in templates.

The property name is now generated via the `IPropertyNameGenerator` provided as part of the `CodeGeneratorSettingsBase` supplied to the `ParameterModelBase` constructor.  This should make them consistent with the component/schema classes in cases where schema references are used.

A new test was added to the `ParameterTests` class in NSwag.CodeGeneration.CSharp.Tests to verify expected behavior.

### CODE REVIEW

File: src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.QueryParameter.liquid [hunk 1]
```diff
@@ -6,16 +6,16 @@ urlBuilder_.Append("{{ parameter.Name }}=").Append(System.Uri.EscapeDataString({
 foreach (var item_ in {{ parameter.VariableName }}) { urlBuilder_.Append("{{ parameter.Name }}=").Append(System.Uri.EscapeDataString(ConvertToString(item_, System.Globalization.CultureInfo.InvariantCulture))).Append("&"); }
 {% elseif parameter.IsDeepObject -%}
 {%     for property in parameter.PropertyNames -%}
-if ({{parameter.Name}}.{{property}} != null)
+if ({{parameter.Name}}.{{property.PropertyName}} != null)
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.QueryParameter.liquid]: I think we should rename PropertyName to Name (same as in the parameter object on the left)

File: src/NSwag.CodeGeneration/Models/ParameterModelBase.cs [hunk 2]
```diff
@@ -44,6 +45,12 @@ public abstract class ParameterModelBase
             Type = typeName;
             Name = parameterName;
             VariableName = variableName;
+
+            var propertyNameGenerator = settings?.PropertyNameGenerator ?? new DefaultPropertyNameGenerator();
```
Comment (RSuter) [inline src/NSwag.CodeGeneration/Models/ParameterModelBase.cs]: Is this (and DefaultPropertyNameGenerator.cs) really needed? Isnt a generator already registered/set by default?

Comment (RSuter) [reply src/NSwag.CodeGeneration/Models/ParameterModelBase.cs]: Yes, I think you can remove DefaultPropertyNameGenerator as there is already one for CSharp and TypeScript which is always set: 

https://github.com/RSuter/NJsonSchema/blob/master/src/NJsonSchema.CodeGeneration.CSharp/CSharpGeneratorSettings.cs#L41

Just throw an exception here, e.g.: 

    settings?.PropertyNameGenerator ?? throw new InvalidOperationException("PropertyNameGenerator not set.");

**Comment by RSuter** [comment]
Thanks for the PR. Looks really good! Is it ready to be merged?

**Comment by sweetlandj** [comment]
Thanks!  Yeah it should be ready to go.  I created packages locally and they seem to behave as expected.

**Pull request merged**